lumicks/pylake/population/mixture.py

```diff
@@ -0,0 +1,115 @@
+import numpy as np
+from sklearn.mixture import GaussianMixture
+from scipy import stats
+
+
+def as_sorted(fcn):
+    """Decorator to return results sorted according to mapping array."""
+
+    def wrapper(self, *args, **kwargs):
+        result = fcn(self, *args, **kwargs)
+        return result[self._map]
+
+    return wrapper
+
+
+class GaussianMixtureModel:
+    def __init__(self, data, K, init_method, n_init, tol, max_iter):
+        """A wrapper around scikit-learn's GMM
+
+        Parameters
+        ----------
+        K : int
+            number of Gaussian components in the model
+        """
+        self.K = K
+        self._model = GaussianMixture(
+            n_components=K, init_params=init_method, n_init=n_init, tol=1e-3, max_iter=100
+        )
+        self._model.fit(data)
+
+    @classmethod
+    def from_slice(cls, slc, K, init_method="kmeans", n_init=1, tol=1e-3, max_iter=100):
+        data = slc.data.reshape((-1, 1))
+        return cls(data, K, init_method="kmeans", n_init=1, tol=1e-3, max_iter=100)
+
+    @property
+    def exit_flag(self):
+        return {
+            "converged": self._model.converged_,
+            "n_iter": self._model.n_iter_,
+            "lower_bound": self._model.lower_bound_,
+        }
+
+    @property
+    def _map(self):
+        """Indices of sorted means."""
+        return np.argsort(self._model.means_.squeeze())
+
+    @property
+    @as_sorted
+    def weights(self):
+        return self._model.weights_
+
+    @property
+    @as_sorted
+    def means(self):
+        return self._model.means_.squeeze()
+
+    @property
+    @as_sorted
+    def variances(self):
+        return self._model.covariances_.squeeze()
+
+    @property
+    def std(self):
+        return np.sqrt(self.variances)
+
+    def label(self, trace):
+        """Label channel trace data as states."""
+        data = trace.data.reshape((-1, 1))
+        labels = self._model.predict(data)  # wrapped model labels
+        output_states = np.argsort(self._map)  # output model state labels in wrapped model order
+        return output_states[labels]  # output model labels
+
+    def bic(self, trace):
+        """Bayesian Information Criterion."""
+        return self._model.bic(trace.data.reshape((-1, 1)))
+
+    def aic(self, trace):
+        """Akaike Information Criterion."""
+        return self._model.aic(trace.data.reshape((-1, 1)))
```

Comment:
Is it useful to compute the AIC/BIC on a _new_ trace? Shouldn't it always use the data you trained with since it's an in-sample method? If not, wouldn't it make sense to just store this thing after training?

Reply:
This functionality is offered by sklearn, but I agree it may not be useful. Default is now to return values only for training data